File: algo_2021/w2.-two-pointer/problems/148.sort_colors/148_sort_colors__two_pass_2.py
class Solution:
    """
    @param nums: A list of integer which is 0, 1 or 2 
    @return: nothing
    """
    
    def sortColors(self, nums):
        # 3 elements require 2 partition 
        self.partition(nums, 1)
        self.partition(nums, 2)
    def partition(self, nums, k):
        last_small_pointer = -1
        for cur in range(len(nums)):
            if nums[cur] < k:
                last_small_pointer+=1
                # swap value
                nums[last_small_pointer], nums[cur] = nums[cur], nums[last_small_pointer]

File: algo_2021/w2.-two-pointer/148.sort_colors/148_sort_colors__two_pass_2.java
public class Solution {
    /**
     * @param nums: A list of integer which is 0, 1 or 2 
     * @return: nothing
     */
    public void sortColors(int[] nums) {
        // write your code here
        partitionArray(nums, 1);
        partitionArray(nums, 2);
    }
    
    private void partitionArray(int[] nums, int k) {
        int lastSmallPointer = -1;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] < k) {
                lastSmallPointer++;
                int tmp = nums[lastSmallPointer];
                nums[lastSmallPointer] = nums[i];
                nums[i] = tmp;
            }
        }
    }
}